crates/ide/src/hover.rs

```diff
@@ -4446,4 +4515,71 @@ fn foo() -> Option<()> {
                 ```"#]],
         );
     }
+
+    #[test]
+    fn hover_deref_expr() {
+        check_hover_range(
+            r#"
+//- minicore: deref
+use core::ops::Deref;
+
+struct DerefExample<T> {
+    value: T
+}
+
+impl<T> Deref for DerefExample<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.value
+    }
+}
+
+fn foo() {
+    let x = DerefExample { value: 0 };
+    let y: i32 = $0*x$0;
+}
+"#,
+            expect![[r#"
+                ```text
+                Type:                            i32
+                Dereferenced from: DerefExample<i32>
+                ```
+            "#]],
+        );
+    }
+
+    #[test]
+    fn hover_deref_expr_with_coercion() {
+        check_hover_range(
+            r#"
+//- minicore: deref
+use core::ops::Deref;
+
+struct DerefExample<T> {
+    value: T
+}
+
+impl<T> Deref for DerefExample<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.value
+    }
+}
+
+fn foo() {
+    let x = DerefExample { value: &&&&&0 };
+    let y: &i32 = $0*x$0;
+}
+"#,
+            expect![[r#"
+                ```text
+                Type:                            &&&&&i32
+                Coerced to:                          &i32
+                Dereferenced from: DerefExample<&&&&&i32>
```

Comment:
Thinking about this, can we make this read something like this instead?
```suggestion
                Dereferenced from: DerefExample<&&&&&i32>
                To Type:                         &&&&&i32
                Coerced to:                          &i32
```
I think this way its easier to figure out what happened.